xds/src/main/java/io/grpc/xds/XdsEndpointResource.java

```diff
@@ -185,21 +191,39 @@ private static int getRatePerMillion(FractionalPercent percent) {
   @VisibleForTesting
   @Nullable
   static StructOrError<LocalityLbEndpoints> parseLocalityLbEndpoints(
-      io.envoyproxy.envoy.config.endpoint.v3.LocalityLbEndpoints proto) {
+      io.envoyproxy.envoy.config.endpoint.v3.LocalityLbEndpoints proto)
+      throws ResourceInvalidException {
     // Filter out localities without or with 0 weight.
     if (!proto.hasLoadBalancingWeight() || proto.getLoadBalancingWeight().getValue() < 1) {
       return null;
     }
     if (proto.getPriority() < 0) {
       return StructOrError.fromError("negative priority");
     }
+
+    ImmutableMap<String, Object> localityMetadata;
+    try {
+      MetadataRegistry registry = MetadataRegistry.getInstance();
+      localityMetadata = registry.parseMetadata(proto.getMetadata());
+    } catch (ResourceInvalidException e) {
+      throw new ResourceInvalidException("Failed to parse Locality Endpoint metadata: "
+          + e.getMessage(), e);
+    }
     List<Endpoints.LbEndpoint> endpoints = new ArrayList<>(proto.getLbEndpointsCount());
     for (io.envoyproxy.envoy.config.endpoint.v3.LbEndpoint endpoint : proto.getLbEndpointsList()) {
       // The endpoint field of each lb_endpoints must be set.
       // Inside of it: the address field must be set.
       if (!endpoint.hasEndpoint() || !endpoint.getEndpoint().hasAddress()) {
         return StructOrError.fromError("LbEndpoint with no endpoint/address");
       }
+      ImmutableMap<String, Object> endpointMetadata;
+      try {
+        MetadataRegistry registry = MetadataRegistry.getInstance();
```

Comment:
Get this outside the for loop and earlier in the method, so you only get it once for the entire method?